### PR TITLE
SolidJS framework feedback

### PR DESCRIPTION
**Pull request by mauriver21**
## Not for merging

This PR is the initial work on a SolidJS framework. As agreed with @JReinhold it is not meant to be merged, it is meant to collect feedback and suggestions. Once that has been done, this PR will be closed and the work will be solidified in a new repo.

### CODE REVIEW

**Comment by JReinhold** [comment]
Marked this as "Ready for Review" just to trigger full CI suite.

File: code/renderers/solid/src/render.tsx [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+  if (force) {
+    flag = true;
+  } else if (!Object.is(globals, context.globals)) {
+    // Globals refers to storybook visualization options.
+    globals = context.globals;
+    flag = true;
+  }
+  return flag;
+};
```
Comment (JReinhold) [inline code/renderers/solid/src/render.tsx:34]: 1. Is the point here to force a re-mount when globals change?
2. Nit: I think the logic can be simplified a little bit.
```suggestion
/**
 * Checks when the story requires to be remounted.
 * Elements outside the story requires a whole re-render.
 * e.g. dark theme, show grid, etc...
 */
const shouldRemount = (force: boolean, context: StoryContext<SolidRenderer>) => {

  if (force) {
    return true;
  }
  
  if (!Object.is(globals, context.globals)) {
    // Globals refers to storybook visualization options.
    globals = context.globals;
    return true;
  }
  
  return false;
};
```

File: code/renderers/solid/src/render.tsx [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+  /**
+   * Store is updated for fine grained updates.
+   */
+  setStore('args', storyContext.args);
+
+  /**
+   * If remount is required, the whole root node is re-rendered.
+   */
+  if (remount(forceRemount, storyContext)) {
+    disposeStory?.();
```
Comment (JReinhold) [inline code/renderers/solid/src/render.tsx:101]: The logic here and the singleton store assumes that only one story will be rendered at once. But in docs mode that is not the case because there we render multiple stories at the same time.
In this case that will mean that they all read and write from the same singleton store above - conflicting - and on this line they'll dispose each other immediately.
You can reproduce this by adding multiple stories and checking the autodocs page, it will only show the last story of the page because the others will have been disposed.

You'd probably need a map of stories being rendered similar to what we have in the Svelte renderer. https://github.com/storybookjs/storybook/blob/next/code/renderers/svelte/src/render.ts#L14
You can key the map by `canvasElement` which will be unique, and I think the value would then be a combination of a store and a dispose function.

Let me know if you need further assistance with this.

You can of course postpone this change, and say that the initial version doesn't support docs mode.

File: code/renderers/solid/template/cli/js/Button.stories.jsx [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+// More on how to set up stories at: https://storybook.js.org/docs/7.0/html/writing-stories/introduction
+export default {
+  title: 'Example/Button',
+  tags: ['autodocs'],
+  render: (args) => <Button {...args}/>,
+};
+
+// More on writing stories with args: https://storybook.js.org/docs/7.0/html/writing-stories/args
+export const Primary = {
+  args: {},
```
Comment (JReinhold) [inline code/renderers/solid/template/cli/js/Button.stories.jsx:12]: There should probably be some content on the button to not confuse people.

```suggestion
  args: { children: 'Click Me' },
```

File: code/renderers/solid/template/cli/ts/Button.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+  tags: ['autodocs'],
+  render: (args) => <Button {...args} />,
+};
+
+export default meta;
+type Story = StoryObj<ButtonProps>;
+
+// More on writing stories with args: https://storybook.js.org/docs/7.0/html/writing-stories/args
+export const Primary: Story = {
+  args: {},
```
Comment (JReinhold) [inline code/renderers/solid/template/cli/ts/Button.stories.tsx:17]:
```suggestion
  args: { children: 'Click Me' },
```

File: code/renderers/solid/template/stories/html-mdx.stories.mdx [hunk 5]
```diff
@@ -0,0 +1,21 @@
+import { Meta, Story, Canvas } from '@storybook/addon-docs';
```
Comment (JReinhold) [inline code/renderers/solid/template/stories/html-mdx.stories.mdx:1]: As stories defined in MDX is being deprecated in 7.0 I suggest you remove this template file.

**Review by JReinhold — COMMENTED** [review]
I did a thorough review, it looks really good!

One thing that I think should be fixed though, is that it doesn't work if stories don't have a `render` property. For some reason it doesn't pick up your exported `render` function from `render.tsx`, and I'm unsure how that is supposed to work exactly.

Maybe @tmeasday has some guidance about that?

**Comment by JReinhold** [comment]
# Changes to `prepareStory`

**tl;dr**
The author has needed to make some changes to how we prepare stories, in order to make args reactive in SolidJS - ie. react to Controls changes. I think it looks good but I'd like someone more experienced like @tmeasday to check too if there's any problems with the proposal.

I've gone back and forth with @webblocksapp on Discord about this, and the way Storybook currently prepare stories means that reactive args gets lost in SolidJS. That's because for Solid to react to args/props changes it converts them all to Proxies behind the scenes, and when we do the argType mapping in `prepareStory` the Proxies are destroyed and args becomes a plain object again.

The proposal here is to change the flow so that the context is prepared after being passed to `renderToCanvas` so that the proxies stay intact. I've tested that argType mapping still works in React, but we should probably test other frameworks like Vue and Angular too though.

There's a chance that this change will also make reactive args in Vue3 possible, something that we currently can't do without re-creating the whole Vue story.

# Splitting up the PR

@webblocksapp we've made som decisions on how we plan on supporting community frameworks like this in the future. We're going to make this monorepo only contain Core frameworks and move the rest out to other separate repos. This means that we also won't have SolidJS in the monorepo.

Long-term (we don't know when yet, but after 7.0 release at least) we'll make it possible for users to run something like `npx sb init --framework storyook-solidjs-vite` that will point to an npm package that then installs the necessary thing.

Short-term we'll accept PRs that changes the CLI here so that we can detect SolidJS projects and install your framework when we detect it.

So my suggestion here would be to close this PR when you want to, and split it up into three things:

1. A separate repo that contains two packages from here: `storybook-solidjs-vite` and `storybook-solidjs` (looks like `storybook-solid` is taken)
1. A PR that contains your changes with the `prepareContext` call.
2. A PR that contains your changes to the CLI so it detects SolidJS and installs `storybook-solidjs-vite`

For the separate repo, it's up to you if you want to create it yourself, or if we should create a repo under this storybookjs org, and you can continue the work there.

**Comment by tmeasday** [comment]
> One thing that I think should be fixed though, is that it doesn't work if stories don't have a render property. For some reason it doesn't pick up your exported render function from render.tsx, and I'm unsure how that is supposed to work exactly.

> Maybe @tmeasday has some guidance about that?

It should be picking it up from the `/preview.js` file, which re-exports `src/config.ts` (<- PS I'd call that `src/preview.ts`, "config" is the old name, we just have updated other renderers).

> The author has needed to make some changes to how we prepare stories, in order to make args reactive in SolidJS - ie. react to Controls changes. I think it looks good but I'd like someone more experienced like @tmeasday to check too if there's any problems with the proposal.

I think I need some more context about this, shall we discuss @JReinhold?

File: code/renderers/solid/src/render.tsx (hunk 1 above)
Comment (webblocksapp) [reply code/renderers/solid/src/render.tsx:34]: 1. Yes because fine-grained updates are not possible in React, so it needs to re-run the component to update the storybook canvas with globals.

File: code/renderers/solid/src/render.tsx (hunk 2 above)
Comment (webblocksapp) [reply code/renderers/solid/src/render.tsx:101]: Thanks for this clarification, this is something that I have pending after having fine-grained updates working.

**Comment by webblocksapp** [comment]
Thanks @JReinhold  for the deep review, I will start working today with your code suggestions.

Related to this:
> One thing that I think should be fixed though, is that it doesn't work if stories don't have a `render` property. For some reason it doesn't pick up your exported `render` function from `render.tsx`, and I'm unsure how that is supposed to work exactly.

I have this story example on my local env. and is working without problems. Could you please provide your failing story to take a look?

```typescript
export const Secondary = {
  args: {
    disabled: true,
    children: "Hello World",
  },
};
```

Related to splitting up the PR:

1. I would like to have the needed separate repos under your storybookjs org, I think we can save work re-migrating again a separate repo from my own. Also I prefer those packages to be under your organization so they will have more visibility, @storybook/solid, @storybook/solid-vite.
2. I will proceed with that once you have confirmed to apply this solution.
3. Got it.

File: code/renderers/solid/src/render.tsx (hunk 1 above)
Comment (tmeasday) [reply code/renderers/solid/src/render.tsx:34]: I don't think this is the "correct" behaviour. Globals should cause a re-render in the same way that args do.

**Comment by tmeasday** [comment]
We discussed the changes preview, where arg type mapping is pulled "up" the rendering chain. The concern was that it is a breaking change in that now decorators will receive mapped/filtered args, whereas in the past they have not.

It seems like that's better behaviour anyway. After surveying a little internally we've decided we are OK with pushing forward with that change. One thing I'd like to see if we are making that change is some unit tests demonstrating the mapping applying to decorators.

**Comment by JReinhold** [comment]
> > One thing that I think should be fixed though, is that it doesn't work if stories don't have a `render` property. For some reason it doesn't pick up your exported `render` function from `render.tsx`, and I'm unsure how that is supposed to work exactly.
> 
> I have this story example on my local env. and is working without problems. Could you please provide your failing story to take a look?

Sorry about that, I interpreted an error wrong. Everything is working as expected, it was just a brainfart on my end.

> I would like to have the needed separate repos under your storybookjs org, I think we can save work re-migrating again a separate repo from my own. Also I prefer those packages to be under your organization so they will have more visibility, @storybook/solid, @storybook/solid-vite.

I'll create a repo for you ASAP. It'll be a monorepo for anything solid related, as the Solid renderer and Solid-Vite framework will probably be developed together.
We won't be releasing them under the `@storybook` npm namespace though for security reasons, but I think `storybook-solidjs` and `storybook-solidjs-vite` is also fine.


The first step towards making the necessary changes to core will be to make a separate PR for them, so we can see tests running. And as Tom said we also want a few unit tests, something that I can probably help out with if you need that.

**Comment by webblocksapp** [comment]
Thanks @JReinhold and @tmeasday for your feedback. 
I will start working with the core change as first step for having args mapped before the render fn.
I hope to have the PR maybe at the end of this week or starting next.

I've accepted too the repo invitation, so I will start the solid framework migration as soon I have PR the core change.

**Comment by webblocksapp** [comment]
Hi @JReinhold and @tmeasday, I've just PRed the core changes. I'm not able to label the PR, so I added it inside the description as a Breaking Change.

I modified 2 unit test files however they are failing due to snapshots. Locally I ran both tests with the -u flag, they are passing however the console threw me 0 snapshots for each file.

I attach PR #20755 and I continue aware of your comments for moving forward with points 1 and 3 or if the current PR needs more changes.

1. A separate repo that contains two packages from here: storybook-solidjs-vite and storybook-solidjs (looks like storybook-solid is taken)
2. ~~A PR that contains your changes with the prepareContext call.~~
3. A PR that contains your changes to the CLI so it detects SolidJS and installs storybook-solidjs-vite
